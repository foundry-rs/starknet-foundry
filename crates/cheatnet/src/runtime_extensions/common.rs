use blockifier::execution::execution_utils::felt_to_stark_felt;
use blockifier::execution::syscalls::hint_processor::SyscallCounter;
use cairo_felt::Felt252;
use cairo_vm::vm::trace::trace_entry::TraceEntry;
use cairo_vm::vm::vm_core::VirtualMachine;
use starknet_api::hash::StarkFelt;
use starknet_api::transaction::Calldata;

pub fn create_execute_calldata(calldata: &[Felt252]) -> Calldata {
    let calldata: Vec<StarkFelt> = calldata.iter().map(felt_to_stark_felt).collect();
    Calldata(calldata.into())
}

#[must_use]
pub fn sum_syscall_counters(mut a: SyscallCounter, b: &SyscallCounter) -> SyscallCounter {
    for (key, value) in b {
        *a.entry(*key).or_default() += *value;
    }
    a
}

#[must_use]
pub fn get_relocated_vm_trace(vm: &VirtualMachine) -> Vec<TraceEntry> {
    vm.get_relocated_trace()
        .unwrap()
        .iter()
        .map(|x| TraceEntry {
            pc: x.pc,
            ap: x.ap,
            fp: x.fp,
        })
        .collect()
}
