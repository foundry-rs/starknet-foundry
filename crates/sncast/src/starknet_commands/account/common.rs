use anyhow::Result;
use std::fs;

use anyhow::Context;
use camino::Utf8PathBuf;
use serde_json::{Value, json};

pub fn load_accounts(accounts_file: &Utf8PathBuf) -> Result<Value> {
    let contents = fs::read_to_string(accounts_file).context("Failed to read accounts file")?;

    if contents.trim().is_empty() {
        return Ok(json!({}));
    }

    let accounts = serde_json::from_str(&contents)
        .with_context(|| format!("Failed to parse accounts file at = {accounts_file}"))?;

    Ok(accounts)
}
