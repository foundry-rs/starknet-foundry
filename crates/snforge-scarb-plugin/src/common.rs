use crate::{
    args::Arguments,
    asserts::assert_is_used_once,
    attributes::AttributeInfo,
    parse::{parse, parse_args},
};
use cairo_lang_macro::{Diagnostic, Diagnostics, ProcMacroResult, TokenStream};
use cairo_lang_syntax::node::{ast::FunctionWithBody, db::SyntaxGroup};
use cairo_lang_utils::Upcast;

#[allow(clippy::needless_pass_by_value)]
pub fn into_proc_macro_result(
    args: TokenStream,
    item: TokenStream,
    handler: impl Fn(&TokenStream, &TokenStream, &mut Vec<Diagnostic>) -> Result<String, Diagnostics>,
) -> ProcMacroResult {
    let mut warns = vec![]; // `Vec<Diagnostic>` instead of `Diagnostics` because `Diagnostics` does not allow to push ready `Diagnostic`

    match handler(&args, &item, &mut warns) {
        Ok(item) => ProcMacroResult::new(TokenStream::new(item)).with_diagnostics(warns.into()),
        Err(diagnostics) => ProcMacroResult::new(item).with_diagnostics(
            //TODO(#2358) extend with warns
            diagnostics,
        ),
    }
}

pub fn with_parsed_values<Collector>(
    args: &TokenStream,
    item: &TokenStream,
    warns: &mut Vec<Diagnostic>,
    handler: impl Fn(
        //func item
        &dyn SyntaxGroup,
        &FunctionWithBody,
        //args
        &dyn SyntaxGroup,
        Arguments,
        //warns
        &mut Vec<Diagnostic>,
    ) -> Result<String, Diagnostics>,
) -> Result<String, Diagnostics>
where
    Collector: AttributeInfo,
{
    let item = item.to_string();
    let (db, func) = parse::<Collector>(&item)?;

    let db = db.upcast();

    assert_is_used_once::<Collector>(db, &func)?;

    let (args_db, args) = parse_args(&args.to_string());
    let args_db = args_db.upcast();

    let args = Arguments::new::<Collector>(args_db, args, warns);

    handler(db, &func, args_db, args, warns)
}
