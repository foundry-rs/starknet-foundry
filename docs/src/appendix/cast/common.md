# `sncast` common flags

## `--profile, -p <PROFILE_NAME>`
Optional.

Profile name in `Scarb.toml` config file.

## `--path-to-scarb-toml, -s <PATH>`
Optional.

Path to `Scarb.toml` file.

If supplied, cast will not look for `Scarb.toml` file in current (or parent) directory, but will use this path instead.

## `--url, -u <RPC_URL>`
Optional.

Starknet RPC node url address.

Overrides url from `Scarb.toml`.

## `--network, -n <NETWORK_NAME>`
Optional.

Starknet network name, one of `testnet`, `testnet2`, `mainnet`.

Overrides network from `Scarb.toml`.

## `--account, -a <ACCOUNT_NAME`
Optional.

Account name used to interact with the network, aliased in open zeppelin accounts file.

Overrides account from `Scarb.toml`.

## `--accounts-file, -f <PATH_TO_ACCOUNTS_FILE>`
Optional.

Path to the open zeppelin accounts file holding accounts info. Defaults to `~/.starknet_accounts/starknet_open_zeppelin_accounts.json`.

## `--int-format, -i`
Optional.

If passed, values will be displayed as decimals instead of hexes.

## `--json, -j`
Optional.

If passed, output will be displayed in json format.

## `--version, -v`

Prints out `cast` version.

## `--help, -h`

Prints out help.
