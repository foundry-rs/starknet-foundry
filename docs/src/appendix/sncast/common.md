# `sncast` common flags

## `--profile, -p <PROFILE_NAME>`
Optional.

Used for both `snfoundry.toml` and `Scarb.toml` if specified.
Defaults to `default` (`snfoundry.toml`) and `dev` (`Scarb.toml`).

## `--url, -u <RPC_URL>`
Optional.

Starknet RPC node url address.

Overrides url from `snfoundry.toml`.

## `--account, -a <ACCOUNT_NAME>`
Optional.

Account name used to interact with the network, aliased in open zeppelin accounts file.

Overrides account from `snfoundry.toml`.

If used with `--keystore`, should be a path to [starkli account JSON file](https://book.starkli.rs/accounts#accounts).

## `--accounts-file, -f <PATH_TO_ACCOUNTS_FILE>`
Optional.

Path to the open zeppelin accounts file holding accounts info. Defaults to `~/.starknet_accounts/starknet_open_zeppelin_accounts.json`.

## `--keystore, -k <PATH_TO_KEYSTORE_FILE>`
Optional.

Path to [keystore file](https://book.starkli.rs/signers#encrypted-keystores).
When specified, the --account argument must be a path to [starkli account JSON file](https://book.starkli.rs/accounts#accounts).

## `--int-format`
Optional.

If passed, values will be displayed in decimal format. Default is addresses as hex and fees as int.

## `--hex-format`
Optional.

If passed, values will be displayed in hex format. Default is addresses as hex and fees as int.

## `--json, -j`
Optional.

If passed, output will be displayed in json format.

## `--wait, -w`
Optional.

If passed, command will wait until transaction is accepted or rejected.

## `--wait-timeout <TIME_IN_SECONDS>`
Optional.

If `--wait` is passed, this will set the time after which `sncast` times out. Defaults to 60s.

## `--wait-retry-timeout <TIME_IN_SECONDS>`
Optional.

If `--wait` is passed, this will set the retry interval - how often `sncast` should fetch tx info from the node. Defaults to 5s.

## `--version, -v`

Prints out `sncast` version.

## `--help, -h`

Prints out help.
