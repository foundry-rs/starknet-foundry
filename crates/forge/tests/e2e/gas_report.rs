use crate::e2e::common::runner::{
    BASE_FILE_PATTERNS, Package, setup_package, setup_package_with_file_patterns, test_runner,
};
use indoc::indoc;
use shared::test_utils::output_assert::assert_stdout_contains;

#[test]
fn basic() {
    let temp = setup_package("simple_package");
    let output = test_runner(&temp)
        .arg("--gas-report")
        .arg("call_and_invoke")
        .assert()
        .code(0);

    assert_stdout_contains(
        output,
        indoc! {r"
    [..]Compiling[..]
    [..]Finished[..]


    Collected 1 test(s) from simple_package package
    Running 0 test(s) from src/
    Running 1 test(s) from tests/
    [PASS] simple_package_integrationtest::contract::call_and_invoke (l1_gas: ~0, l1_data_gas: ~[..], l2_gas: ~[..])
    ╭------------------------+-------+-------+-------+---------+---------╮
    | HelloStarknet Contract |       |       |       |         |         |
    +====================================================================+
    | Function Name          | Min   | Max   | Avg   | Std Dev | # Calls |
    |------------------------+-------+-------+-------+---------+---------|
    | get_balance            | 13340 | 13340 | 13340 | 0       | 2       |
    |------------------------+-------+-------+-------+---------+---------|
    | increase_balance       | 24940 | 24940 | 24940 | 0       | 1       |
    ╰------------------------+-------+-------+-------+---------+---------╯

    Tests: 1 passed, 0 failed, 0 ignored, [..] filtered out
    "},
    );
}

#[test]
fn recursive_calls() {
    let temp = setup_package("debugging");

    let output = test_runner(&temp)
        .arg("test_debugging_trace_success")
        .arg("--gas-report")
        .assert()
        .code(0);

    assert_stdout_contains(
        output,
        indoc! {r"
    [..]Compiling[..]
    [..]Finished[..]


    Collected 1 test(s) from debugging package
    Running 0 test(s) from src/
    Running 1 test(s) from tests/
    [PASS] debugging_integrationtest::test_trace::test_debugging_trace_success (l1_gas: ~0, l1_data_gas: ~[..], l2_gas: ~[..])
    ╭-------------------------+-------+--------+--------+---------+---------╮
    | SimpleContract Contract |       |        |        |         |         |
    +=======================================================================+
    | Function Name           | Min   | Max    | Avg    | Std Dev | # Calls |
    |-------------------------+-------+--------+--------+---------+---------|
    | execute_calls           | 11680 | 609080 | 183924 | 235859  | 5       |
    |-------------------------+-------+--------+--------+---------+---------|
    | fail                    | 17950 | 17950  | 17950  | 0       | 1       |
    ╰-------------------------+-------+--------+--------+---------+---------╯

    Tests: 1 passed, 0 failed, 0 ignored, [..] filtered out
    "},
    );
}

#[test]
fn multiple_contracts_and_constructor() {
    let temp = setup_package("simple_package_with_cheats");
    let output = test_runner(&temp)
        .arg("--gas-report")
        .arg("call_and_invoke_proxy")
        .assert()
        .code(0);

    assert_stdout_contains(
        output,
        indoc! {r"
    [..]Compiling[..]
    [..]Finished[..]


    Collected 1 test(s) from simple_package_with_cheats package
    Running 0 test(s) from src/
    Running 1 test(s) from tests/
    [PASS] simple_package_with_cheats_integrationtest::contract::call_and_invoke_proxy (l1_gas: ~0, l1_data_gas: ~[..], l2_gas: ~[..])
    ╭------------------------+-------+-------+-------+---------+---------╮
    | HelloStarknet Contract |       |       |       |         |         |
    +====================================================================+
    | Function Name          | Min   | Max   | Avg   | Std Dev | # Calls |
    |------------------------+-------+-------+-------+---------+---------|
    | get_block_number       | 15780 | 15780 | 15780 | 0       | 2       |
    ╰------------------------+-------+-------+-------+---------+---------╯

    ╭-----------------------------+--------+--------+--------+---------+---------╮
    | HelloStarknetProxy Contract |        |        |        |         |         |
    +============================================================================+
    | Function Name               | Min    | Max    | Avg    | Std Dev | # Calls |
    |-----------------------------+--------+--------+--------+---------+---------|
    | constructor                 | 14650  | 14650  | 14650  | 0       | 1       |
    |-----------------------------+--------+--------+--------+---------+---------|
    | get_block_number            | 125280 | 125280 | 125280 | 0       | 2       |
    ╰-----------------------------+--------+--------+--------+---------+---------╯

    Tests: 1 passed, 0 failed, 0 ignored, [..] filtered out
    "},
    );
}

#[test]
fn fork() {
    let temp =
        setup_package_with_file_patterns(Package::Name("forking".to_string()), BASE_FILE_PATTERNS);

    let output = test_runner(&temp)
        .arg("test_track_resources")
        .arg("--gas-report")
        .assert()
        .code(0);

    assert_stdout_contains(
        output,
        indoc! {r"
        [..]Compiling[..]
        [..]Finished[..]

        Collected 1 test(s) from forking package
        Running 1 test(s) from src/
        [PASS] forking::tests::test_track_resources (l1_gas: ~0, l1_data_gas: ~[..], l2_gas: ~[..])
        ╭---------------------------+-------+-------+-------+---------+---------╮
        | forked contract           |       |       |       |         |         |
        | (class hash: 0x06a7…1550) |       |       |       |         |         |
        +=======================================================================+
        | Function Name             | Min   | Max   | Avg   | Std Dev | # Calls |
        |---------------------------+-------+-------+-------+---------+---------|
        | get_balance               | 40000 | 40000 | 40000 | 0       | 1       |
        |---------------------------+-------+-------+-------+---------+---------|
        | increase_balance          | 40000 | 40000 | 40000 | 0       | 1       |
        ╰---------------------------+-------+-------+-------+---------+---------╯
        ╭---------------------------+-------+-------+-------+---------+---------╮
        | forked contract           |       |       |       |         |         |
        | (class hash: 0x07aa…af4b) |       |       |       |         |         |
        +=======================================================================+
        | Function Name             | Min   | Max   | Avg   | Std Dev | # Calls |
        |---------------------------+-------+-------+-------+---------+---------|
        | get_balance               | 13840 | 13840 | 13840 | 0       | 1       |
        |---------------------------+-------+-------+-------+---------+---------|
        | increase_balance          | 25840 | 25840 | 25840 | 0       | 1       |
        ╰---------------------------+-------+-------+-------+---------+---------╯

        Tests: 1 passed, 0 failed, 0 ignored, [..] filtered out
        "},
    );
}

#[test]
fn no_transactions() {
    let temp = setup_package("simple_package");
    let output = test_runner(&temp)
        .arg("--gas-report")
        .arg("test_fib")
        .assert()
        .code(0);

    assert_stdout_contains(
        output,
        indoc! {r"
    [..]Compiling[..]
    [..]Finished[..]


    Collected 1 test(s) from simple_package package
    Running 1 test(s) from src/
    [PASS] simple_package::tests::test_fib (l1_gas: ~0, l1_data_gas: ~0, l2_gas: ~[..])
    No contract gas usage data to display. Make sure your test include transactions.

    Running 0 test(s) from tests/
    Tests: 1 passed, 0 failed, 0 ignored, [..] filtered out
    "},
    );
}
