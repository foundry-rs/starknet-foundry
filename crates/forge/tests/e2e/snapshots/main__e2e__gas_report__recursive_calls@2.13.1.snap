---
source: crates/forge/tests/e2e/gas_report.rs
expression: stdout
---
[PASS] debugging_integrationtest::test_trace::test_debugging_trace_success (l1_gas: ~0, l1_data_gas: ~288, l2_gas: ~1372230)
╭-------------------------+-------+--------+--------+---------+---------╮
| SimpleContract Contract |       |        |        |         |         |
+=======================================================================+
| Function Name           | Min   | Max    | Avg    | Std Dev | # Calls |
|-------------------------+-------+--------+--------+---------+---------|
| execute_calls           | 11660 | 609380 | 184000 | 235987  | 5       |
|-------------------------+-------+--------+--------+---------+---------|
| fail                    | 17950 | 17950  | 17950  | 0       | 1       |
╰-------------------------+-------+--------+--------+---------+---------╯
